datafusion/src/physical_plan/mod.rs

```diff
@@ -509,6 +540,43 @@ pub trait Accumulator: Send + Sync + Debug {
     fn evaluate(&self) -> Result<ScalarValue>;
 }
 
+/// A window accumulator represents a stateful object that lives throughout the evaluation of multiple
+/// rows and generically accumulates values.
+///
+/// An accumulator knows how to:
+/// * update its state from inputs via `update`
+/// * convert its internal state to a vector of scalar values
+/// * update its state from multiple accumulators' states via `merge`
+/// * compute the final value from its internal state via `evaluate`
+pub trait WindowAccumulator: Send + Sync + Debug {
+    /// scans the accumulator's state from a vector of scalars, similar to Accumulator it also
+    /// optionally generates values.
+    fn scan(&mut self, values: &[ScalarValue]) -> Result<Option<ScalarValue>>;
+
+    /// scans the accumulator's state from a vector of arrays.
+    fn scan_batch(&mut self, values: &[ArrayRef]) -> Result<Option<Vec<ScalarValue>>> {
```

Comment:
Following up on @Dandandan 's comment https://github.com/apache/arrow-datafusion/pull/380#discussion_r638136080 since Arrow is column based already, creating `Vec`s of `ScalarValues` seems like it may be limiting performance

What would you think about using  [slice](https://docs.rs/arrow/4.1.0/arrow/array/trait.Array.html#method.slice) instead to pick subsets of arrays (rather than `Vec<ScalarValue>`)?

I realize this code is following the model of `Accumulator` so I think it would also be ok to continue in this style and we can go back and rewrite in terms of `Array` later if needed

Reply:
it is used down the road, in [1]. I think if possible i can change the return type of `scan_batch` directly to `ArrayRef` so that upon scanning multiple batches, a single `concat` is needed.

[1]: https://github.com/apache/arrow-datafusion/pull/413/files#diff-8b6b5ea3976c91229244e4e7a31a7026422b1374d1683e44b41af67a6bd43187R432

Reply:
@alamb I have updated to use `ArrayRef` instead of `Vec<ScalarValue>` whereas possible